gio/src/list_model.rs

```diff
@@ -19,15 +31,149 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
         }
         res
     }
+    fn iter<LT: IsA<glib::Object>>(&self) -> Result<ListModelIter<LT>, &Self>
+    where
+        Self: Sized + Clone,
+    {
+        if !self.item_type().is_a(LT::static_type()) {
+            return Err(self);
```

Comment:
Would be nicer to return a proper error type here. Returning `Self` only really makes sense for functions that take it by value

Reply:
True, but this keeps `iter` similar to `downcast`. This type check feels similar to what downcast is doing. I know downcast returns `Err(Self)`... I try to return the most similar thing I can

Reply:
`downcast_ref()` returns an `Option` :) The only reason why `downcast()` returns `Self` in the error case is because it takes `Self` by value and it would otherwise be gone on error.

Reply:
> The only reason why downcast() returns Self in the error case is because it takes Self by value and it would otherwise be gone on error.

I hadn't thought of that 😅